In [1]:
# Importing required libraries
import pandas as pd
import numpy as np

In [34]:
movies_df = pd.read_csv('F:\\Learning\\Python Projects\\Recommender system tutorial\\ml-latest-small\\movies.csv')

In [35]:
rating_df=pd.read_csv('F:\\Learning\\Python Projects\\Recommender system tutorial\\ml-latest-small\\ratings.csv')

In [36]:
movies_df.shape

(9742, 3)

In [37]:
rating_df.shape

(100836, 4)

In [38]:
movies_df.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

We are buulding a content based recommender system. So our main focus will be in the properties of the content (movies). Here the property that we have is 'genres'. Our aim here is to split the genre in individual category and create a new feature for each individual genre.

For doing this we will have to follows these steps:
    1. identify all unique genres
    2. create empty columns with each genre in the dataframe
    3. populate the columns with 1 where-ever applicable

In [88]:
# creating a list of genres which will have list of genres for each movie
list_of_list_of_genres = []
for t in movies_df.genres:
    list_of_list_of_genres.append(t.split('|'))

In [89]:
len(list_of_list_of_genres)

9742

The length of newly created list is as same as that of the database

In [92]:
list_of_list_of_genres[0:5]

[['Adventure', 'Animation', 'Children', 'Comedy', 'Fantasy'],
 ['Adventure', 'Children', 'Fantasy'],
 ['Comedy', 'Romance'],
 ['Comedy', 'Drama', 'Romance'],
 ['Comedy']]

In [93]:
def flatten(test_list):
    if isinstance(test_list, list):
        temp = []
        for ele in test_list:
            temp.extend(flatten(ele))
        return temp
    else:
        return [test_list]

In [94]:
flattened_list = flatten(list_of_list_of_genres)

In [96]:
unique_genres  = (set(flattened_list))

In [98]:
len(unique_genres)

20

there are 20 unique genres, meaning there will be 20 columns created in the database

In [47]:
# creating columns by genre name with initial values as 0
for c in unique_genres:
    movies_df[c] = 0

In [48]:
# putting 1 inplace of 0 under each column where ever applicable for indivudual movies
for i in movies_df.index:
    for c in movies_df.columns[3:23]:
        for l in list_of_list_of_genres[i]:
            if l == c:
                #print('True')
                movies_df.at[i,c] = 1
            #else:
            #    movies_df.at[i,c] = 0

In [99]:
movies_df.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  Comedy  Fantasy  Western  \
0  Adventure|Animation|Children|Comedy|Fantasy       1        1        0   
1                   Adventure|Children|Fantasy       0        1        0   
2                               Comedy|Romance       1        0        0   
3                         Comedy|Drama|Romance       1        0        0   
4                                       Comedy       1        0        0   

   Action  Drama  Crime  Animation  ...  Thriller  Musical  Romance  IMAX  \
0       0      0      0          1  ...         0        0        0     0   
1       0      0      0          0  ...         0        0        0     0   
2       0      0      0          0  ...         0        0        1     0   
3       0      1      0          0  ...         0        0        1     0   
4       0      0      0          0  ...         0        0        0     0   

   Mystery  Sci-Fi  Documentary  Film-Noir  Children  Adventure  
0        0       0            0          0         1          1  
1        0       0            0          0         1          1  
2        0       0            0          0         0          0  
3        0       0            0          0         0          0  
4        0       0            0          0         0          0  

[5 rows x 23 columns]

In [49]:
movies_df.loc[0]

movieId                                                         1
title                                            Toy Story (1995)
genres                Adventure|Animation|Children|Comedy|Fantasy
Comedy                                                          1
Fantasy                                                         1
Western                                                         0
Action                                                          0
Drama                                                           0
Crime                                                           0
Animation                                                       1
(no genres listed)                                              0
Horror                                                          0
War                                                             0
Thriller                                                        0
Musical                                                         0
Romance   

In [51]:
movies_df.iloc[:,3:23]

Comedy  Fantasy  Western  Action  Drama  Crime  Animation  \
0          1        1        0       0      0      0          1   
1          0        1        0       0      0      0          0   
2          1        0        0       0      0      0          0   
3          1        0        0       0      1      0          0   
4          1        0        0       0      0      0          0   
...      ...      ...      ...     ...    ...    ...        ...   
9737       1        1        0       1      0      0          1   
9738       1        1        0       0      0      0          1   
9739       0        0        0       0      1      0          0   
9740       0        0        0       1      0      0          1   
9741       1        0        0       0      0      0          0   

      (no genres listed)  Horror  War  Thriller  Musical  Romance  IMAX  \
0                      0       0    0         0        0        0     0   
1                      0       0    0         0        0        0     0   
2                      0       0    0         0        0        1     0   
3                      0       0    0         0        0        1     0   
4                      0       0    0         0        0        0     0   
...                  ...     ...  ...       ...      ...      ...   ...   
9737                   0       0    0         0        0        0     0   
9738                   0       0    0         0        0        0     0   
9739                   0       0    0         0        0        0     0   
9740                   0       0    0         0        0        0     0   
9741                   0       0    0         0        0        0     0   

      Mystery  Sci-Fi  Documentary  Film-Noir  Children  Adventure  
0           0       0            0          0         1          1  
1           0       0            0          0         1          1  
2           0       0            0          0         0          0  
3           0       0            0          0         0          0  
4           0       0            0          0         0          0  
...       ...     ...          ...        ...       ...        ...  
9737        0       0            0          0         0          0  
9738        0       0            0          0         0          0  
9739        0       0            0          0         0          0  
9740        0       0            0          0         0          0  
9741        0       0            0          0         0          0  

[9742 rows x 20 columns]

In [55]:
# importing required libraries
from sklearn.metrics.pairwise import cosine_similarity

In [102]:
# Compute the cosine similarity matrix
cosine_sim = cosine_similarity(X = movies_df.iloc[:,3:23])

In [103]:
cosine_sim

array([[1.        , 0.77459667, 0.31622777, ..., 0.        , 0.31622777,
        0.4472136 ],
       [0.77459667, 1.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.31622777, 0.        , 1.        , ..., 0.        , 0.        ,
        0.70710678],
       ...,
       [0.        , 0.        , 0.        , ..., 1.        , 0.        ,
        0.        ],
       [0.31622777, 0.        , 0.        , ..., 0.        , 1.        ,
        0.        ],
       [0.4472136 , 0.        , 0.70710678, ..., 0.        , 0.        ,
        1.        ]])

In [104]:
cosine_sim.shape

(9742, 9742)

as we can see, the cosine similarity matrix shape is 9742 x 9742. So if we take the first row from this as shown below, it will mean that we have xtracted similarity score generated for Toy Story (since Toy Story is the first movie in our database)

In [105]:
cosine_sim[0]

array([1.        , 0.77459667, 0.31622777, ..., 0.        , 0.31622777,
       0.4472136 ])

and if we again extract the first value from this array it will give the similarity score of Toy Story with Toy Story which will be 1. If we extract the second value it will give us similarity score of Toy Story with Jumanji

In [106]:
cosine_sim[0][0] # similarity of Toy Story with Toy Story

0.9999999999999999

In [107]:
cosine_sim[0][1] # similarity of Toy Story with Jumanji

0.7745966692414835

In [60]:
#Construct a reverse map of indices and movie titles
movie_indices = pd.Series(movies_df.index, index=movies_df['title']).drop_duplicates()

In [61]:
movie_indices

title
Toy Story (1995)                                0
Jumanji (1995)                                  1
Grumpier Old Men (1995)                         2
Waiting to Exhale (1995)                        3
Father of the Bride Part II (1995)              4
                                             ... 
Black Butler: Book of the Atlantic (2017)    9737
No Game No Life: Zero (2017)                 9738
Flint (2017)                                 9739
Bungo Stray Dogs: Dead Apple (2018)          9740
Andrew Dice Clay: Dice Rules (1991)          9741
Length: 9742, dtype: int64

In [76]:
index_position_of_basemovie = indices['Jumanji (1995)']

In [77]:
cosine_sim[index_position_of_basemovie]

array([0.77459667, 1.        , 0.        , ..., 0.        , 0.        ,
       0.        ])

In [79]:
sim_scores = list(enumerate(cosine_sim[index_position_of_basemovie]))

In [80]:
sim_scores_sorted = sorted(sim_scores,
                           key=lambda x: x[1], # to sort based on similarity scores
                           reverse=True)

In [82]:
sim_scores_top10 = sim_scores_sorted[1:11]

In [83]:
sim_scores_top10

[(53, 1.0000000000000002),
 (109, 1.0000000000000002),
 (767, 1.0000000000000002),
 (1514, 1.0000000000000002),
 (1556, 1.0000000000000002),
 (1617, 1.0000000000000002),
 (1618, 1.0000000000000002),
 (1799, 1.0000000000000002),
 (3574, 1.0000000000000002),
 (6075, 1.0000000000000002)]

In [109]:
movie_indices_top10 = [i[0] for i in sim_scores_top10]

In [110]:
movie_indices_top10

[53, 109, 767, 1514, 1556, 1617, 1618, 1799, 3574, 6075]

In [111]:
movies_df['title'].iloc[movie_indices_top10]

53                     Indian in the Cupboard, The (1995)
109                     NeverEnding Story III, The (1994)
767                       Escape to Witch Mountain (1975)
1514            Darby O'Gill and the Little People (1959)
1556                                  Return to Oz (1985)
1617                        NeverEnding Story, The (1984)
1618    NeverEnding Story II: The Next Chapter, The (1...
1799                        Santa Claus: The Movie (1985)
3574    Harry Potter and the Sorcerer's Stone (a.k.a. ...
6075    Chronicles of Narnia: The Lion, the Witch and ...
Name: title, dtype: object

In [113]:
# checking if the similarities are really true
movies_df.iloc[movie_indices_top10,0:3]

movieId                                              title  \
53         60                 Indian in the Cupboard, The (1995)   
109       126                  NeverEnding Story III, The (1994)   
767      1009                    Escape to Witch Mountain (1975)   
1514     2043          Darby O'Gill and the Little People (1959)   
1556     2093                                Return to Oz (1985)   
1617     2161                      NeverEnding Story, The (1984)   
1618     2162  NeverEnding Story II: The Next Chapter, The (1...   
1799     2399                      Santa Claus: The Movie (1985)   
3574     4896  Harry Potter and the Sorcerer's Stone (a.k.a. ...   
6075    41566  Chronicles of Narnia: The Lion, the Witch and ...   

                          genres  
53    Adventure|Children|Fantasy  
109   Adventure|Children|Fantasy  
767   Adventure|Children|Fantasy  
1514  Adventure|Children|Fantasy  
1556  Adventure|Children|Fantasy  
1617  Adventure|Children|Fantasy  
1618  Adventure|Children|Fantasy  
1799  Adventure|Children|Fantasy  
3574  Adventure|Children|Fantasy  
6075  Adventure|Children|Fantasy